### PR TITLE
Add double run protection to operation

### LINKED ISSUE #29: Calling execute! twice...executes twice

### PR DESCRIPTION
**Pull request by garside**
This closes #29

### CODE REVIEW

File: lib/flow/operation/errors/already_executed.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+module Operation
+  module Errors
+    class AlreadyExecuted < StandardError; end
+  end
+end
```
Comment (arettberg) [inline lib/flow/operation/errors/already_executed.rb]: hrmm...does this really need its own file? I think convention is to have a single `namespace/errors` file. Remember- we're in gem land, autoloading isn't a thing

File: spec/operation/error_handler_spec.rb [hunk 2]
```diff
@@ -1,7 +1,12 @@
 # frozen_string_literal: true
 
 RSpec.describe Operation::ErrorHandler, type: :module do
-  include_context "with an example operation", [ Operation::Failures, Operation::Execute, described_class ]
+  include_context "with an example operation", [
+    Operation::Failures,
+    Operation::Execute,
+    Operation::Status,
+    described_class,
+  ]
```
Comment (arettberg) [inline spec/operation/error_handler_spec.rb]: I once rejected this formatting, then was forced with the same decision in my own code and did the same thing 😇

File: spec/operation/execute_spec.rb [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
   it { is_expected.to include_module ActiveSupport::Rescuable }
 
-  include_context "with an example operation", Operation::Execute
+  include_context "with an example operation", [ described_class, Operation::Status ]
```
Comment (arettberg) [inline spec/operation/execute_spec.rb]: it's not abundantly clear why this happened...

File: spec/operation/execute_spec.rb [hunk 4]
```diff
@@ -92,6 +92,14 @@
         end
       end
     end
+
+    context "when called twice" do
+      before { example_operation.execute! }
+
+      it "raises" do
+        expect { execute! }.to raise_error Operation::Errors::AlreadyExecuted
```
Comment (arettberg) [inline spec/operation/execute_spec.rb]: can we maybe assert that whatever action _would_ have been taken by the operation wasn't called?

File: spec/operation/error_handler_spec.rb (hunk 2 above)
Comment (garside) [reply spec/operation/error_handler_spec.rb]: whitespace strategy: Force majeure

File: lib/flow/operation/errors/already_executed.rb (hunk 1 above)
Comment (garside) [reply lib/flow/operation/errors/already_executed.rb]: Yeah I'm explicitly including it.

Do you mean a master common errors file for flow? I have them namespaced under each module of flow at the moment.

I would suggest if you want to refactor it, it's a perfect time to standardize now that there's a second example of this. 

Easy way to get into the contributors file 😎 

tl;dr: agreed, but lazy; you wanna do it? 😬

File: spec/operation/execute_spec.rb (hunk 3 above)
Comment (garside) [reply spec/operation/execute_spec.rb]: Well, I added an internal; dependency to the `::Execute` module for a `executed?` method, which is provided by the `Status` module.

The rename to `described_class` is because RSpec standards are hard to keep 100% pure if you intend to 😅

File: spec/operation/execute_spec.rb (hunk 4 above)
Comment (garside) [reply spec/operation/execute_spec.rb]: ohyeah!

File: lib/flow/operation/errors/already_rewound.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+module Operation
+  module Errors
+    class AlreadyRewound < StandardError; end
+  end
+end
```
Comment (garside) [inline lib/flow/operation/errors/already_rewound.rb]: Sorry! 😓 

`#justfollowingconvention`
